Firestore/core/src/firebase/firestore/core/event_listener.h

```diff
@@ -0,0 +1,141 @@
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_EVENT_LISTENER_H_
+
+#include <atomic>
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/executor.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor_callback.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+/**
+ * A general interface for listening to events internally.
+ */
+template <typename T>
+class EventListener {
+ public:
+  static std::unique_ptr<EventListener<T>> Create(
+      util::StatusOrCallback<T>&& callback);
+
+  virtual ~EventListener() {
+  }
+
+  /**
+   * OnEvent will be called with the new value or the error if an error
+   * occurred.
+   *
+   * @param maybe_value The value of the event or the error.
+   */
+  virtual void OnEvent(util::StatusOr<T> maybe_value) = 0;
+};
+
+/**
+ * A wrapper around another EventListener that dispatches events asynchronously.
+ */
+template <typename T>
+class AsyncEventListener
+    : public EventListener<T>,
+      public std::enable_shared_from_this<AsyncEventListener<T>> {
+ public:
+  using DelegateListener = std::unique_ptr<EventListener<T>>;
+
+  AsyncEventListener(util::Executor* executor, DelegateListener&& delegate)
+      : executor_(executor), delegate_(std::move(delegate)) {
+  }
+
+  static std::shared_ptr<AsyncEventListener<T>> Create(
+      util::Executor* executor, DelegateListener&& delegate);
+
+  static std::shared_ptr<AsyncEventListener<T>> Create(
+      util::Executor* executor, EventListener<T>&& delegate) {
+    return Create(executor,
+                  absl::make_unique<EventListener>(std::move(delegate)));
+  }
+
+  void OnEvent(util::StatusOr<T> maybe_value) override;
+
+  /**
+   * Synchronously mutes the listener and raise no further events. This method
+   * is thread safe can be called from any queue.
+   */
+  void Mute();
+
+ private:
+  std::atomic<bool> muted_{false};
+  util::Executor* executor_;
+  DelegateListener delegate_;
+};
+
+template <typename T>
+std::unique_ptr<EventListener<T>> EventListener<T>::Create(
+    util::StatusOrCallback<T>&& callback) {
+  class CallbackEventListener : public EventListener<T> {
+   public:
+    explicit CallbackEventListener(util::StatusOrCallback<T>&& callback)
+        : callback_(std::move(callback)) {
+    }
+
+    void OnEvent(util::StatusOr<T> maybe_value) override {
+      callback_(std::move(maybe_value));
+    }
+
+   private:
+    util::StatusOrCallback<T> callback_;
+  };
+
+  return absl::make_unique<CallbackEventListener>(std::move(callback));
+}
+
+template <typename T>
+std::shared_ptr<AsyncEventListener<T>> AsyncEventListener<T>::Create(
+    util::Executor* executor, DelegateListener&& delegate) {
+  return std::make_shared<AsyncEventListener<T>>(executor, std::move(delegate));
+}
+
+template <typename T>
+void AsyncEventListener<T>::Mute() {
+  muted_.store(true);
```

Comment:
I had used `load`/`store` initially because I dislike hiding important operations behind operators. However, I can see the appeal. I've changed it because you pointed it out.

Reply:
I understand the rationale, though in this case an argument can be made that any operation on an atomic variable is atomic, so as long as the reader of the code keeps the type in mind, there should be no ambiguity.